File: 1962-single-threaded-cpu/single-threaded-cpu.java
class Solution {

    class Task
    {
        int index;
        int processingTime;
        int startTime;
        Task(int startTime,int processingTime,int index)
        {
            this.processingTime=processingTime;
            this.startTime=startTime;
            this.index=index;
        }
    }

    public int[] getOrder(int[][] tasks) 
    {   
        int n=tasks.length;
        Task [] tasksList=new Task[n];

        for(int i=0;i<n;i++)
        tasksList[i]=new Task(tasks[i][0],tasks[i][1],i);

        Arrays.sort(tasksList,
            (a,b)-> {
                // Compare based on start time
                return a.startTime-b.startTime;
                // // Compare based on processing time if start time is same
                // return a.processingTime-b.processingTime;
            }
        );

        PriorityQueue<Task> minHeap = new PriorityQueue<>(
            (a,b) -> {
                // Compare based on the processing time
                if(a.processingTime!=b.processingTime)
                return a.processingTime-b.processingTime;
                // Compare based on the index if processing time is same
                return a.index-b.index;
            }
        );

        int []order=new int[n];
        int i=0,j=0;
        int time=0;
        while(i<n || !minHeap.isEmpty())
        {
            if(minHeap.isEmpty() && time<tasksList[i].startTime)
            time=tasksList[i].startTime;

            while(i<n && tasksList[i].startTime<=time)
            {
                minHeap.add(tasksList[i]);
                i++;
            }

            Task t=minHeap.poll();
            order[j++]=t.index;
            time+=t.processingTime;
        }
        return order;
    }
}

File: 1962-single-threaded-cpu/single-threaded-cpu.cpp
class Solution {
public:

    class Task {
        public:

        int index;
        int pTime;
        int startTime;
        Task(int startTime,int pTime,int index)
        {
            this->index=index;
            this->pTime=pTime;
            this->startTime=startTime;
        }
    };

    class myComparator {
    public:

        bool operator()(Task a,Task b)
        {
            // Min-heap based on processing time
            if(a.pTime!=b.pTime)
            return a.pTime>b.pTime;
            // If pTime is the same, sort by index
            return a.index>b.index;

            // priority_queue mein ULTA LIKHNA PADTA HAI.
        }
    };
    

    vector<int> getOrder(vector<vector<int>>& tasks) 
    {
        int n=tasks.size();
        vector<Task> tasksList;
        for(int i=0;i<n;i++)
        tasksList.push_back(Task(tasks[i][0],tasks[i][1],i));

        // sorting tasksList based on increasing startTime.
        sort( tasksList.begin(), tasksList.end(), [](Task a,Task b){return a.startTime<b.startTime;});

        vector<int> order;
        priority_queue<Task,vector<Task>,myComparator> pq;
        int i=0;
        long time=0;
        while(i<n || !pq.empty())
        {
            if(pq.empty() && time<tasksList[i].startTime)
            time=tasksList[i].startTime;

            while(i<n && time>=tasksList[i].startTime)
            {
                pq.push(tasksList[i]);
                i++;
            }

            Task t=pq.top();
            pq.pop();

            order.push_back(t.index);
            time+=t.pTime;
        }
        return order;
    }
};